File: src/beatsync/extrapolation.ts
import { PhaseSample, BarPhaseSample } from "./schema";

export function predictPhase(sample: PhaseSample, currentTime: number): number {
  const timeSinceUpdate = currentTime - sample.lastUpdateTime;
  const ibi = 60.0 / sample.bpm;
  const phaseAdvance = (timeSinceUpdate / ibi) * 2 * Math.PI;
  let predictedPhase = sample.phase + phaseAdvance;
  return Math.atan2(Math.sin(predictedPhase), Math.cos(predictedPhase));
}

export function predictBarPhase(sample: BarPhaseSample, currentTime: number): number {
  const timeSinceUpdate = currentTime - sample.lastUpdateTime;
  const ibi = 60.0 / sample.bpm;
  const beatsAdvanced = timeSinceUpdate / ibi;
  return sample.barPhase + beatsAdvanced;
}

File: src/beatsync/wasm/extrapolation.cpp
#include "extrapolation.h"
#include <cmath>

double predictPhase(const StabilizerTrajectory& trajectory, double currentTime) {
    double timeSinceUpdate = currentTime - trajectory.lastUpdateTime;
    double ibi = 60.0 / trajectory.bpm;
    double phaseAdvance = (timeSinceUpdate / ibi) * 2 * M_PI;
    double predictedPhase = trajectory.phase + phaseAdvance;
    return std::atan2(std::sin(predictedPhase), std::cos(predictedPhase));
}

double predictBarPhase(const StabilizerTrajectory& trajectory, double currentTime) {
    double timeSinceUpdate = currentTime - trajectory.lastUpdateTime;
    double ibi = 60.0 / trajectory.bpm;
    double beatsAdvanced = timeSinceUpdate / ibi;
    return trajectory.barPhase + beatsAdvanced;
}
